tests/loggen/loggen_plugin.c

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2007-2018 Balabit
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include <stdio.h>
+#include <sys/types.h>
+#include <stdlib.h>
+#include <errno.h>
+#include <sys/time.h>
+#include <time.h>
+
+#include "loggen_plugin.h"
+#include "loggen.h"
+#include "loggen_helper.h"
+
+gboolean
+thread_check_exit_criteria(thread_data *thread_context)
+{
+  if (thread_context->option->number_of_messages != 0
+      && thread_context->sent_messages >= thread_context->option->number_of_messages)
+    {
+      DEBUG("(thread %d) number of sent messages reached the defined limit (%d)\n",thread_context->index,
+            thread_context->option->number_of_messages);
+      return TRUE;
+    }
+
+  struct timeval now;
+  gettimeofday(&now, NULL);
+
+  if ( time_val_diff_in_sec(&now, &thread_context->start_time) > thread_context->option->interval
```

Comment:
Can these two stop conditions run in parallel? If not, it would worth splitting the function, and branching based on which kind of stop condition to check. As I recall, gettimeofday can be slow, lets calling if not necessary.

Reply:
I swap the condition's order:
if ( !thread_context->option->permanent &&
      time_val_diff_in_sec(&now, &thread_context->start_time) > thread_context->option->interval )
Now the gettimeofday will be called if we are running without permanent option. In this case we really need to check the actual time even if it is slow.